Pxb 2.4 5.7.40 merge 2990

VarunNagaraju opened the pull request.

bkandasa committed:
Raise version number after cloning 5.6.53

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

bkandasa committed:
Raise version number after cloning 5.7.35

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

Commit:
Bug #32919524 	CONTRIBUTION: BACKPORT BUG#28949700: SUBOPTIMAL USE OF STRING::RESERVE() IN ...

Contributed by Annirudh Prasad

Backport of BUG#28949700 SUBOPTIMAL USE OF STRING::RESERVE() IN
WRAPPER_TO_STRING().

The one-argument version of String::reserve() increases the capacity
of the String object linearly and causes a high number of memory
allocations when converting certain JSON documents to text
representation.

This patch changes wrapper_to_string() so that it uses the
two-argument version of String::reserve() instead, which enables
exponential increase of the capacity and reduces the number of memory
allocations.

The patch includes some additional small improvements to make the
conversion from JSON to text go faster:

- When appending a string literal, provide the length as an argument
  to avoid calls to strlen().

- When appending two-character string literals, instead append the two
  characters individually (append(char) is inlined, append(const
  char *, size_t) is not).

Change-Id: I5ee36791d07c6525918b2ae0e5919ce47f05404e

Commit:
Bug #32998981 JSON-DOM-T BROKEN 5.7 BUILD WITH GCC 8.4.1

After upgrading to gcc 8.4.1 on el8 we hit an internal compiler error:

In file included from unittest/gunit/merge_large_tests.cc:20:
unittest/gunit/json_dom-t.cc: In member function
virtual void json_dom_unittest::JsonDomTest_BasicTest_Test::TestBody():
unittest/gunit/json_dom-t.cc:392:33:
internal compiler error: in pop_local_binding, at cp/name-lookup.c:2047
   EXPECT_EQ(null_dom, dom.get());

The fix is to split the BasidTest in two.

Change-Id: I57f8c71d954a1015272bd63301e1eb4ab73694f7

marcalff committed:
BUG#31761802 STATISTICS ANY QUERIES USING VIEWS ARE SUMMARIZED TOGETHER WITH THE VIEW DEFINITION SELECT

This patch is for 5.7 only,
backport of 8.0 fix.

Problem:
========

Assuming a view defined as:
  (a) CREATE VIEW v1 AS SELECT * from t1;

And assuming a query like:
  (b1) SELECT col_a, col_b FROM v1;
  (b2) SELECT col_c FROM v1 where ...;

The query digest computed for the queries (b1), (b2)
is based on the view definition (a), instead of the query (b1), (b2).

This is a problem, because:
- several different queries using the same view are all
  computed with the same digest, and all aggregated together
  in table performance_schema.events_statements_summary_by_digest.
- this makes statistics in events_statements_summary_by_digest
  useless when a view is involved.

Root cause:
===========

When executing a query like (b1), the parser is invoked,
tokens from the query ('SELECT', 'col_a', ...) are recorded
in the query digest, and a digest is computed.
So far the result is correct and as expected.

But later during the statement execution,
the server opens the view definition (a),
and makes another call to the parser,
to find out how to open the view itself.

This nested call computes a statement digest again,
this time based on the view definition,
and this digest is recorded with the top level query.

The final result is that the query digest is computed correctly,
then overridden with something irrelevant,
causing the bug.

Solution:
=========

Added a new attribute in the parser:
  bool Parser_input::m_has_digest
to clarify when a digest is to be computed.

Make sure in parse_view_definition()
to not compute digest when invoking the parser
on the view definition.

Consequences:
=============

Before this fix, queries (b1) and (b2) had the same digest,
and were considered to be the "same" query.

After this fix, the digests of (b1) and (b2) are different,
and they no longer collide into the same value.

Other tools that rely on query digests, such as:
- the query rewrite plugin
- the firewall plugin
will now see (b1) and (b2) as different queries,
which is the correct result.

Existing rules for the query rewrite plugin and the firewall plugin
will need to be updated to treat (b1) and (b2) correctly.

Approved by: Chris Powers <chris.powers@oracle.com>

Commit:
Bug#32771259: PAGE COMPRESSION OF INNODB TABLE IS IGNORED DURING RECOVERY PROCESS AND MIGHT NEVER SUCCESS WITH ENOSPC CRASH NEAR TO DISK-FULL

space->compression_type is set only at open of the table with the dictionary.
But during recovery process, the dictionary is not recovered yet also.

The problem is all flushing pages dring recovery never compressed. (because space->compression_type is not set yet)
And flushing causes increasing size of the .ibd file and more possibility to cause disk-full crash of the recovery process...

We need to estimate space->compression_type. (Not 100% exact, but much better than always false)
And it can be done with the second page (page_no:1) flag, when get_file_size reads page_no:0.

Not to increase IO opportunity, reads the page with header page (page_no:0) at Fil_shard::get_file_size() only if recv_recovery_is_on().

* This fix is the first-aid treatment. The true solution is recovered consistent flag in the DD at the time.

RB: 26269
Reviewed-by: Debarun Banerjee <debarun.banerjee@oracle.com>

sachinagarwal1111 committed:
Bug #32831765 SERVER HITS OOM CONDITION WHEN LOADING TWO INNODB
TABLES WITH FTS INDEXES

Problem:
Concurrent INSERTs into the tables (with FTS index) causes huge
number of FTS_SYNC requests are sent and rapid growth in memory
usages.
When concurrent inserts into single table, FTS_OPT thread get SYNC
request when fts cache threshold is reached. In this case FTS_OPT
thread fetches the SYNC request and performs FTS SYNC operation.
When concurrent inserts into many tables, SYNC requests are sent for
many tables. At a time, FTS_OPT thread can process one SYNC request.
In the meantime, each INSERT into other table keeps sending SYNC
request to FTS_OPT thread.

Fix:
1. added a variable total_size_before_sync in fts cache to record total
size of fts cache when SYNC request is sent.
2. next SYNC request is sent based on the delta fts cache size
increment from last SYNC request.
3. reset total_size_before_sync when SYNC completes.

RB: #26571
Reviewed by : Rahul Agarkar <rahul.agarkar@oracle.com>

marcalff committed:
BUG#31761802 STATISTICS ANY QUERIES USING VIEWS ARE SUMMARIZED TOGETHER WITH THE VIEW DEFINITION SELECT

Test cleanup for 5.7 embedded.

Approved by: Chris Powers <chris.powers@oracle.com>

Commit:
BUG#32965864 MULTI-THREADED UNSAFE ACCESS TO A FUNCTION LOCAL STATIC VARIABLE

Description
-----------
Both `MYSQL_BIN_LOG::report_missing_purged_gtids` and
`MYSQL_BIN_LOG::report_missing_gtids` have an output parameter name
`errmsg` that will, under given conditions, be instantiated with a
non-thread-local function-local static variable reference.

Invocation of both methods may occur from different threads execution
stacks and concurrent changes to the static variable may occur.

Side effects include threads logging an error message intended for other
threads or a scrambled error message.

Fix
---
Use `std::string` as the output parameter type, instead of `const char**`
and remove the `static` modifier from the function-local variable.

Reviewed-by: Justin Jose <justin.jose@oracle.com>
Reviewed-by: Pedro Gomes <pedro.gomes@oracle.com>
RB: 26731

Commit:
Bug#32843447: GTID_PURGED IS GETTING EMPTY AND GTID_EXECUTED
              IS GETTING UPDATED AFTER RESTARTING MYSQL

Description: The state of the server in regards to GTIDs is
shared between the binary logs and the `mysql.gtid_executed`
table and both are needed in different scenarios to keep the
consistency of the state.

Analysis: Issue in the reported case is the following:
1. The dump returned by `mysqldump` includes the
   `SET GTID_PURGED` set to the value of `GTID_EXECUTED` and
   the dropping of the `mysql` schema, in that order.
2. `RESET MASTER` clears the server state (GTID state
    included), meaning there are no binary logs.
3. The restore of the dump does the following, in the
   described order:
   a. Executes `SET GTID_PURGED`, which writes the GTID
      state to the `mysql.gtid_executed` table.
   b. It drops the `mysql` schema, which contains the
      `gtid_executed` table. Since the produced dump file
      doesn't restore the contents of the
      `mysql.gtid_executed` table, the changes in 3.a. get
      undone.
Since there are no binary logs, the data added to
`mysql.gtid_executed` was the only persisted value of the
GTID state, at the point just after `SET GTID_PURGED`
execution.

Fix : The issue is fixed in 2 parts :
1. Added a new option, '--skip-mysql-schema', in mysqldump
which when set should not allow mysql schema to be dropped.
2. Updated the sequence in which the `GTID_PURGED` is set
in the dump file.

RB#26690

Commit:
Bug#33162828: INNODB: Assertion failure: ut0ut.cc:552 thread 140294520874752

Problem:
Server crashes when an undo tablespace is truncated while
an active transaction is using it.

Solution:
Do not mark the transaction as complete until all the cleanup is done.

RB# 26771

Reviewed By: Debarun Banerjee <debarun.banerjee@oracle.com>

Commit:
BUG#33181859: Followup for Bug 32800020 - the deadlocking continues

Problem :
---------
In bug#32800020 bug fix, we fixed the deadlock by truncating pending
rollback segments (rseg) during undo tablespace truncate. This patch
attempts to resolve the issue by freeing old rsegs during undo space
truncate. It works for anyone upgrading directly to the patched version
from 5.6. It fails to correct the old rseg if the issue has already
occurred.

1. We could face the same deadlock till the time a space is truncated.

2. Once server is restarted after hang old rseg could have references
   to open transaction undo belonging to new rseg. This would cause
   assert while truncating undo tablespace.

Solution :
----------
1. During server start-up, find rouge 5.6 (pre 5.7.2) rsegs and reset
   the slot. Emit NOTE in mysqld error log.

   [Note] InnoDB: Found duplicate reference rseg: 33 space: 1 page: 3
   [Note] InnoDB: Reset pre-5.7.2 rseg: 1 after duplicate is found.

2. If there is no undo data left to purge don't allocate old_rseg and
   reset the slot. Emit NOTE in mysqld error log.

   [Note] InnoDB: Successfully reset 32 pre-5.7.2 rseg slots.

3. If there is any undo data found for pre-5.7.2 rseg slot, allocate
   old rseg. Emit NOTE recommending slow shutdown in mysqld error log.

   [Note] InnoDB: pre-5.7.2 rseg: 2 holds data to be purged.
          History length: 1. Recommend slow shutdown with
          innodb_fast_shutdown=0 and restart

Reviewed-by: Sachin Z Agarwal <sachin.z.agarwal@oracle.com>

RB: 26810

salmankhan123417 committed:
BUG#31716706 BAD HANDLING OF QUOTES IN SHOW GRANTS

Description: SHOW GRANTS quotes username with ' and if the username
contains ' it is not escaped. When the output of SHOW GRANTS is
dumped to a file to re-excecute the commands will result in either
invalid SQL or SQL injection.

Analysis: It is hardcoded to quote the username with ' and also fails
to escaping the ' when present in username. Right way is to append the
quotes and escape it with \ when quotes are present in username or host.

Fix: Escape the ' with \ if present in username@host specification.

RB: 26845

zmur committed:
Bug#33284372 MySQL 5.7 (and Cluster 7.4, 7.5, 7.6) build fails on macos with XCode

Hide bzero macro defined in system headers.

Change-Id: I5132bbee491817d801de96dff6f8fdb0dcc19338

Commit:
Bug#32552332 - CAN'T WRITE; DUPLICATE KEY IN TABLE '/TMP/#SQL...'

In the case of a query that uses temporary table for aggregation,
the group-by Item is used as the unique constraint of the temporary
table.

If the group-by Item value is already present, the row is updated.
Otherwise, a new row is inserted into the temporary table.

If the group-by Item is an Item with a result field(like Item_func)
or a reference Item, it gets evaluated twice. Once to check if the
result exists in the temporary table and if not, again while
constructing the row to insert.

If the group-by item is non-deterministic, the result value used to
check for existence would be different from the value with which an
insert is tried. This causes the insert to fail if the value already
exists in the table.

Fix:

If any of the group-by Items are non-deterministic, use the hash of
the group-by Items as the unique constraint of the temporary table as
the hash gets evaluated only once.

Change-Id: I7004f298dc0ea84d760903c30af8f08e072a3f2d

zmur committed:
Bug#33284372 MySQL 5.7 (and Cluster 7.4, 7.5, 7.6) build fails on macos with XCode

XCode includes VERSION file from MySQL source code root instead of C++
language header <version>.

Using MySQL source code root as include path is only used in
storage/perfschema, removing that usage make build succeed.

Change-Id: If7d97dca9ebd14068e7eff1c3171812b4075b9ec

sachinagarwal1111 committed:
Bug #33053297 VIRTUAL INDEX CORRUPTED DURING CASCADE UPDATE ON CHILD
TABLE

Problem:
Child table has FOREIGN KEY (FK) constraint with referential action
ON DELETE SET NULL (or ON UPDATE SET NULL).  And child table also
has an index over (FK column, virtual column).  When a row is
deleted from parent table, FK constraint check on child table, set
both [FK column, virtual column] value in the index to NULL.

Fix:
Instead of setting virtual column value to NULL, derives it from the
base columns.  when FK column is set to NULL in the child table,
three cases need to be taken care while setting virtual column value
in the index:
1. If all base columns are in FK then set virtual column value
to NULL.
2. If no base columns is in FK then set virtual column value to
OLD value.
3. If one or more base columns are in FK then set those base
column value to NULL and compute virtual column value again.

RB: 26751
Reviewed by : Rahul Agarkar <rahul.agarkar@oracle.com>

bkandasa committed:
Raise version number after cloning 5.6.54

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

bkandasa committed:
Update License Book

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

bkandasa committed:
Raise version number after cloning 5.7.36

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

wikigurram committed:
Bug#29268656: BOGUS CONSTANT, RANDOM OR TIMEZONE-DEPENDENT
              EXPRESSIONS IN (SUB)PARTITIONING FU

Description
===========
Creating a table with non-deterministic functions in the
generated column expression, should not be allowed as per
the documentation, but this was not being enforced.

Additionally, creating such a table with partitions whose
expression did not actually refer to the problematic column,
failed with a misleading error message about the partition
expression.

Analysis
========
During create table statement parsing, if there are any
unsafe expressions defined for a generated column then
thd->safe_to_cache_query will be set to false. This variable
is later checked while parsing the partition expression.
This check will report an error even if the partition
expression does not refer to the problematic generated
column expression.

Fix
===
Reset the thd->safe_to_cache_query value before parsing the
partition function.

Change-Id: I56e62cd3553db3773b8d0144555ac9a7660224c0

Commit:
Bug #29882299: ALTER USER ... IDENTIFIED WITH ... BY ... SHOULD BE A PRIVILEGED
               OPERATION

Problem: user when tries to change plugin, CREATE USER privilege is not checked.
Fix: Check if IDENTIFIED WITH clause is specified or not when connected user
changes the plugin for his own account. Note: this is backport of bug27923149

RB#27015

phulakun committed:
Bug#33141958 - THE FIRST ASAN UAF ISSUE OF MYSQL SERVER

Statements which commits a transaction implicitly or explicitly are
not allowed inside a trigger and a stored function. CREATE TRIGGER
and CREATE FUNCTION reports an error in this case.

DROP TABLESPACE statement implicitly commits a transaction. But
while creating a trigger and a stored function, this statement is
allowed by mistake. Because of which, assertion to make sure
statement is allowed inside a trigger or a stored function, fails
while executing it in the debug build. An error is reported in the
non-debug build.

To fix this issue, included DROP TABLESPACE statement in the
list of statements which are not allowed inside a trigger and a
stored function. CREATE TRIGGER and CREATE FUNCTION reports an
error if body contains DROP TABLESPACE statement.

Change-Id: I18b89ae4b63e7b1b40d5b3961a7b9eb9711f849c

Commit:
Bug#33238010: partition table access second index lead to cpu high load

The mysql_template of m_prebuilt (and some flags) seems always same among each partitions during same query.
build_template() calls are 1 at first of query is optimal.

RB: 26729
Reviewed-by: Debarun Banerjee <debarun.banerjee@oracle.com>

Commit:
Bug#33417058: The fuction buf_validate() affects performance seriously and can be optimized

NOTE: Performance problem only on debug build.

buf_pool_validate_instance() is called for each BP instances.
But it calls buf_LRU_validate(), all BP instance validation. (for all instances)
It should call buf_LRU_validate_instance().

RB: 27073
Reviewed-by: Marcin Babij <marcin.babij@oracle.com>

Commit:
Bug #33468647 Backport cmake code from Bug#30359287 and Bug#28170711 to 5.7

Comments in cmake/ssl.cmake for 5.7 say:

  For Windows or macOS, WITH_SSL="system" is handled a bit differently:
  We assume you have installed
      https://slproweb.com/products/Win32OpenSSL.html
      We look for "C:/Program Files/OpenSSL-Win64/"
      The .dll files must be in your PATH.

Backport cmake code from
    Bug #30359287 CMAKE THINKS PERL IS SYSTEM SSL ON WINDOWS
    Bug #28170711 ADD CMAKE SUPPORT FOR OPENSSL 1.1 ON WINDOWS

This allows us to build with headers and libraries
from slproweb, ignoring files from  e.g. Strawberry Perl.

Look for static libraries, so that we don't have to extend PATH.

Change-Id: I45ffb0a26074484c02dccf5cccb231266125cd24

pedrolgomes committed:
BUG#33213841: MySQL crashes while executing `RESET REPLICA ALL`

When the replica SQL mode includes the mode 'PAD_CHAR_TO_FULL_LENGTH',
the channel name can be read from the repository with trailing spaces
what can lead to errors in operations that use that name to identify
the channel.

To solve this issue when reading from replication repo table we unset
the mode "PAD_CHAR_TO_FULL_LENGTH"

ReviewBoard: 27047
ReviewBoard: 27008

Commit:
Bug#33141966 - INCONSISTENT BEHAVIOR IF A COLUMN OF TYPE SERIAL IS SET
AS GENERATED

A column of type SERIAL expands to BIGINT UNSIGNED NOT NULL
AUTO_INCREMENT UNIQUE. Defining a column as both SERIAL and GENERATED
ALWAYS causes undefined behaviour.

Fix:
Do not allow a column of type SERIAL to be a generated column.

Change-Id: I4299726e891d7609f5cfc2dc6d1298fa53d9b0b8

karthik-kamath committed:
BUG#28142052: BACKPORT FIX FOR BUG 24293015 TO 5.7

The problem was that SHOW PROCESSLIST could end up reading
freed memory when accessing THD::m_query_string belonging to
a connection that was in the process of deleting a
Prepared_statment (e.g. due to reprepare).

The reason was that THD::m_query_string could be allocated
on the wrong MEM_ROOT - not the statement MEM_ROOT, but the
MEM_ROOT belonging to the prepared statement. So that even
if SHOW PROCESSLIST locks the mutex protecting
THD::m_query_string, the prepared statement could make the
string invalid by deleting
Prepared_statement::main_mem_root.

This patch fixes the problem by making sure that
THD::m_query_string is allocated on the correct memroot
before the prepared statement terminates.

The patch also includes some minor refactoring related to
'expanded_query'.

Note:
This is a backport of the patch for Bug#24293015 from 8.0 to
5.7.

Change-Id: Idd3a074d932afd5e672898490847f83a76236300

salmankhan123417 committed:
BUG#32803050 REPLICA WITH FEDERATED TABLE CRASHES IN HA_FEDERATED::RND_POS WHEN APPLYING ROW EVENTS

Description: When executing Update/Delete with indexes on a table with
Row based replication setup and the federated table on slave, The Slave
crashes.

Analysis: Slave crashes when trying to get the reference to the row
to be update or deleted on federated table using position() method.
The crash occurs due to assert(stored_result) in position method, since
the stored_result is still empty. Server should do index_scan and
update the stored_result and then call for position() and ha_rnd_pos()
methods to get the ref to the row stored. But in this scenario the
server calls for position() method without scanning the table resulting
in crash at assert.

Fix: The fix is added to fetch the position or reference of the row
only when stored_result is not empty.

RB: 26755

gkodinov committed:
Bug #31933415: OUT OF BOUND READ IN ER()
RB#27195

Fixed the case of unknown error.

Commit:
Bug#32989716: MISSING CORRECT CLOSE SESSION HANDLING DURING
              KEYRING MIGRATION

Description: Keyring migration server uses a mysql
connection to perform online migration. This mysql session
is not closed gracefully after the migration. This leads to
a 'Note' being printed on the online server error log:-
"[Note] Aborted connection x to db: 'xyz' user: 'abc' host:
'localhost' (Got an error reading communication packets)"

Analysis: After the online keyring migration, migration
server process is aborted before gracefully closing the
mysql session. This happens because the migration server
process exits before calling the destructor of
'Migrate_keyring', where the closing of the mysql session
is handled.

Fix: This issue is fixed in 8.0 as part of a code
refactoring done in this area by WL#13446. Back-porting the
relevant code changes to 5.7 to fix this bug and to make
the code consistent across versions.

RB#27186

Commit:
Bug #33362907 mysql crashes due to small thread_stack value

Add a check_stack_overrun() call in PT_option_value_list::contextualize(),
since we may call super::contextualize() too late.

Change-Id: If92ff658ca47c755f08e2b67e3e665d3ec7ce2ae
(cherry picked from commit db62015b75d60b8bb4b6f7235a51423c899de123)

Commit:
Bug#31374305 - FORMAT() NOT DISPLAYING WHOLE NUMBER SIDE CORRECTLY FOR
ES_MX AND ES_ES LOCALES

Changed the grouping and decimal separator for spanish locales as per
ICU.

Change-Id: I5d80fa59d3e66372d904e17c22c532d4dd2c565b

Commit:
Bug #33568767 Upgrade to libedit 20210910-3.1 [patch 1]

Patch 1: git add the original tarball.

https://thrysoee.dk/editline/libedit-20210910-3.1.tar.gz

Remove before checkin: config.guess config.sub, m4/lt~obsolete.m4
m4/lt-obsolete.m4 contains illecal characters '~'

Change-Id: Ifcc1a110699f372cebeb7d2689359cc1bb5ef6e3
(cherry picked from commit 0a84fde76c48c64031c8eb4873327344b94368b6)
(cherry picked from commit 11a943ec744a53c76b2e8189e545ae0580f2c71d)

wikigurram committed:
Bug#32975221: 5.7.29+ REGRESSION, SPATIAL INDEX NOT USED
              ANYMORE FOR MBRCONTAINS

Description:
============
This is a regression from the fix for Bug#29770705. The fix
introduced cloning of range tree for a special case when we
have a predicate having an AND condition ("pred1 AND pred2")
and one part of this AND condition is not a simple
predicate(an OR condition - "pred2"). For such a case, we
needed to clone the range tree created for "pred1". However,
when we clone the range tree, we do not copy the "geometry
flag" associated with the range tree. This results in
ignoring the spatial index available to do the range scan.

Fix
===
Copy the `rkey_func_flag` from original tree if GEOM_FLAG is
set.

On 8.0, this issue is not reproducible because the use_count
handling is different and clone() is not invoked. However it
is clear that for some case it might fail. So we need the
changes, hence added the unit test for the same.

Change-Id: I36312bc2c11db4e4e77d67e9443a98b37ade38e8

Commit:
Bug #33568767 Upgrade to libedit 20210910-3.1 [patch 2]

Patch 2: Add cmake code based on the .ac and .am autotools files.

No relevant changes to .ac and .am files, so copy
  src/CMakeLists.txt
  src/config.h

Rename src/makelist to src/makelist.in so we can subst @AWK@ in it.
Change path to current libedit in cmake/readline.cmake.

Change-Id: I4f4c25a2194a2f434f4e1ec5680ca4116d77c4d9
(cherry picked from commit 7c962babc74fab81c07c091606b5be64d2fdc36a)
(cherry picked from commit 86c5543769f198c80d98e35eb7b94abdfda23193)

Commit:
Bug #33568767 Upgrade to libedit 20210910-3.1 [patch 3]

Patch 3:

Add patch for:
Bug #31360025 UPDATE TO LIBEDIT 20190324-3.1 BREAKS CONTROL-C BEHAVIOUR IN MYSQL CLIENT
handle SIGINT in read_char()

No need to apply patch for:
Bug#11757855 49967: BUILT-IN LIBEDIT DOESN'T READ .EDITRC ON LINUX
upstream ifdefs have been cleaned up.

Change-Id: Id1ae08ff213be0bc3c5d957b4cf2840fd00aee07
(cherry picked from commit 660a8d1c29592c70f3a2fef65cad66490bde3b28)
(cherry picked from commit 741abce3ac26b4f049712921657d91ec00f5509b)

Commit:
Bug #33568767 Upgrade to libedit 20210910-3.1 [patch 4]

Patch 4:

Remove old version: extra/libedit/libedit-20191231-3.1

Change-Id: Ied89826b8c74c2e1d47c9e3287577a0fef6d0b0d
(cherry picked from commit 31a598a1acfe7818350daa8f2ff07e8cf2371cbf)
(cherry picked from commit 293d2a8c03cde51a0f7a7914829eeeb0e43366f2)

Commit:
Bug #33568767 Upgrade to libedit 20210910-3.1 [Solaris]

Additional patch, fixing build break on Solaris.

Solaris has MIN in <utility.h> and <sys/sysmacros.h>
(rather than <sys/param.h>)

Change-Id: I1e319e6d580758097e495f58fd1fa7bc8e91a08b
(cherry picked from commit 798b9a207238b3218e506666efd07d168c11fd05)

bkandasa committed:
Updated copyright year in user visible text

    Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

bkandasa committed:
Updated copyright year in user visible text

    Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

bkandasa committed:
Raise version number after cloning 5.6.55

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

bkandasa committed:
Update License Book

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

bkandasa committed:
Raise version number after cloning 5.7.37

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

sachinagarwal1111 committed:
Bug #33101844 CREATE FULLTEXT INDEX CRASHES SERVER FOR LARGE TABLE

Problem:
When an Index is added to the table. Index entries are
first sorted & merged and then index tree is created in bottom up
order to expedite the processing. while creating index tree,
X-latch is taken on the rightmost page at each level and kept till
the end of index creation. meanwhile if page cleaner thread chooses
any of the non-leaf page (or root page) from index tree for flush
then page cleaner thread keeps waiting for X-latch to be released
from that page and eventually hits fatal timeout error.

Fix:
During index creation, X-latch is taken on the non-leaf page
only for insert into that page and release immediately.

RB: 27333
Reviewed by : Annamalai Gurusami <annamalai.gurusami@oracle.com>

MRajcic committed:
Bug#33619511: MySQL server exits in debug build when -p parameter used

Fixes server exit due to uninitialized variables in keyring migration code.

Approved by: Harin Vadodaria <harin.vadodaria@oracle.com>

Change-Id: I05bf3dfd059140c4c1ac0e14a9260e51f5a40ca8

nacarvalho committed:
BUG#33581529: Backport Fix for Bug 29223844 to 5.7 version of Group Replication

An OOM can happen on transaction marshalling to the network in the
group_replication_trans_before_commit() hook.

To solve the above issue that situation is handled resulting in the
transaction rollback.

RB: 27439

MRajcic committed:
Bug#33619511: MySQL server exits in debug build when -p parameter used

post-push fix: Fix for mysql-5.7 using C++03 standard

Change-Id: I6efb2974a4222a7fcd0423bc16ae510357f5f215

MRajcic committed:
Bug#33619511: MySQL server exits in debug build when -p parameter used

post-push fix: Removing test in MySQL 5.7 only, it behaves differently compared to 8.0

Change-Id: I401e6ecf07d13b03abe09815a9cc22c9464cf6e6

harinvadodaria committed:
Bug#33578113: DROP privilege on performance_schema.* can't be revoked

Description: Just like GRANT on PFS and tables within
             it is allowed, REVOKE should be allowed
             too.

Fix: Updated PFS_internal_schema_access::check() to
     allow privilege checks to proceed as usual for
     SQLCOM_REVOKE when privilege in question is
     DROP_ACL.

RB: 27429

Commit:
Bug #33659435 Bundled libevent: EVENT__HAVE_WORKING_KQUEUE not defined with Xcode 12 or later

Apply upstream patch: CheckWorkingKqueue.cmake: fix missing headers

Avoid implicitly declaring functions memset() and exit()
as it is considered an error by some compilers
(e.g. Xcode clang 12 and later)

Change-Id: I6d302ec4c4d08d86e91ebadf78007ad05d58112b
(cherry picked from commit 277b62d0933ef925ffdf499c77735abbc0972fb3)
(cherry picked from commit c1e96e8dffb760834722547a4ea9cfbc35d3d5ec)

Commit:
Bug #33668341 MySQL 5.7: Look for openssl@1.1 on Mac

Homebrew has added support for OpenSSL 3.0
This is currently not supported by our codebase,
so look explicitly for openssl 1.1

Change-Id: I3856cf1830ab14d32317f864e4b0122dcb2371c1

salmankhan123417 committed:
BUG#30042589 MYSQLDUMP LEAKS MEMORY WHEN SELECTED TABLES ARE DUMPED WITH --ORDER-BY-PRIMARY

Description: Executing mysqldump on more than 1 table with
--order-by-primary option leaks memory.

Analysis: Dump_table() is invoked once for each table when executing
dump_selected_tables, So memory for global order_by is allocated
every time. But while freeing its done only once and the previously
allocated memory gets leaked. Memory for order_by is allocated when
get_table_structure() is being called.

Fix: Order_by is freed within dump_table() thus freeing it for every
invocation and also its freed after every get_table_structure() call.

RB: 27465

gkodinov committed:
WL#14937: disable multi-threaded myisam repair in 8.0
RB#27440

The 5.7 version.

Commit:
Bug#33334507 - 5.7 - utf8mb4 + REPLACE(UUID(),'-','') returns
duplicate UUIDs!

Item_func_uuid gets improperly cached and hence returns the same
value for subsequent calls due to missing flag.

Fix:
Set RAND_TABLE_BIT flag to Item_func_uuid so as to be considered
non-constant and further on: non-cacheable.

Change-Id: I772aa15e91d08e5590b24af4bfcc2c6d2428157e

Commit:
Bug #34017614 Add support for Apple M1 to MySQL 5.7 [33688395]

Backport aa93982a34c1851bdc8d738bebdc6dbf9da8040f
Bug #33688395 macos12 ARM warning: 'SIZE_MAX' macro redefined [-Wmacro-redefined]

Use SIZE_MAX from <stdint.h>

Change-Id: I3a9234f42e8e54953ff85aff58054c3fa26b5132
(cherry picked from commit aa93982a34c1851bdc8d738bebdc6dbf9da8040f)

Commit:
BUG #33459653 : coredump after add index for partion table with ibd
file missing
Problem:
- If we delete .ibd file of a partition along with the 1st partition
  then after restart when we open the partition table we get the
  error message, which is the expected behaviour.
- When we delete .ibd file of a partition other than the 1st
  partition then after restart when we open the partition table we
  hit the debug assert because ha_innopart::open() checks for
  ibd_file_missing only for the 1st partition.

Solution:
Instead of checking ibd_file_missing only for the 1st
partition we check ibd_file_missing for all the partitions.

Reviewed by : Marcin Babij (marcin.babij@oracle.com)
RB  #27494

Commit:
Post push fix

Reviewed by : Marcin Babij (marcin.babij@oracle.com)
RB  #27494

sachinagarwal1111 committed:
Bug #33600109 Segmentation fault at ha_innobase::index_type，
Improper code handling

Problem:
The server may terminates abruptly because nullptr check is
missing for index instance.  The index object is returned by
innobase_get_index() and the return value may be 'nullptr'.
This means that the input parameter of the function
dict_index_is_spatial() is nullptr. That will cause the server
either hits a debug assert or reports a segmentation fault.

Fix:
Added a nullptr check for index instance.

RB: 27517
Reviewed by : Debarun Banerjee <debarun.banerjee@oracle.com>

gkodinov committed:
WL#14938: Remove multi-threaded myisam repair
RB#27618

the 5.7 version

nacarvalho committed:
WL#15060: Change group_replication_transaction_size_limit default in 5.7

The option `group_replication_transaction_size_limit` limits the
size of transactions allowed in a Group Replication primary, that
is, if a given transaction exceeds this value it will be rollback.
This option was added in 5.7.19 and 8.0.2-dmr, since 5.7 was a GA
version it was added with a default value of 0 to avoid behaviour
change. On 8.0 the default value is 150000000.

When its value is 0 the memory used for transactions is unbounded,
which can cause out of memory errors.
To prevent these errors we did change the 5.7 default to
150000000, the already default value on 8.0.

RB: 27623

Commit:
Bug 26906787 - INCORRECT AUTO_INCREMENT IDS ISSUED WHEN CLOSE TO ULLONG_MAX

PROBLEM
-------
1) ha_innobase::get_auto_increment() is supposed to return the previous
   auto inc value if   the increment causes overflow ( i.e maximum size of the column)
2) We were not recognizing the maximum value of the column and proceeding
   to calculate the auto inc value which causes wrong values to be generated.
FIX
---
1) Recognize if the increment causes overflow and return the previous auto inc value

#rb 27654
Reviewed-by : Debarun Banerjee <debarun.banerjee@oracle.com>

trosten committed:
Bug#33903639 Fix spell of log messages in mysql_safe

 Fix some spelling mistakes in mysqld_safe.
 Based on contribution from bin wang.

Change-Id: Ia4a7e99f3b42d754dd7db6c5470381e41ec9d368

Commit:
BUG #32586721 : SERVER CRASH IN FIL_SPACE_IS_BEING_TRUNCATED

Problem:
- If we start server with keyring and delete record from encrypted table,
  keeping innodb_purge_stop_now=ON.
- On restarting the server, purge threads will try to process undo records
  of encrypted table. However tablespace for encrypted table will not be
  loaded.
- This will result in termination of the server.

Fix:
  Skip undo records for which tablespace is missing.

Reviewed by : Annamalai Gurusami (annamalai.gurusami@oracle.com)
RB  #27559

Commit:
Bug#33732907: Slow query log logs password in plain text on syntax error

If a statement could not be parsed, it would not be included in
the general log (unless --log-raw was used, in which case the
statement would be logged before parsing).

By contrast, the slow query log would include such a query,
which in theory could expose sensitive credentials.

This patch changes the behaviour so that unparseable queries are
not slow-logged. This aligns it with the default behaviour of
the general log. The slow query log will not observe --log-raw
or align its behaviour with the performance schema's, as it is
not the correct tool to debug malformed queries or hunt down
possible attempts at SQL injection.

Backport from 8.x

Change-Id: I2395bc853b8c66a60ab6e1578dceffbb48de2c05

Commit:
Bug#27492990 	PARALLEL REPLICATION ALWAYS FAILS WITH SPECIFIC WORKLOAD FROM SYSBENCH

Background

Change-Id: I74bc82fcdfaae28e2c005b46b93084b45653fbb8
----------
If we want to apply binary log files as relay log files in a server that has
the same server id as the id in the files then we can do it using
replicate_same_server_id flag.

Problem
-------
The replica applier coordinator fails to adapt its internal transaction
boundary trackers in the presence of a Rotate event when
replicate_same_server_id is enabled and the event has the same server id.

Solution
--------
In this case, we should make sure that the value of replicate_same_server_id
is properly checked in the conditions in
`Rotate_log_event::do_update_pos(Relay_log_info *rli)`.

Change-Id: I74bc82fcdfaae28e2c005b46b93084b45653fbb8

bkandasa committed:
Raise version number after cloning 5.6.56

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

bkandasa committed:
Update License Book

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

bkandasa committed:
Raise version number after cloning 5.7.38

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

marcalff committed:
WL#14611 BACKPORT SHOW PROCESSLIST TO 5.7

This patch is for 5.7, to merge to 8.0

Reserving a new error number in all releases:
  ER_WARN_WRONG_NATIVE_TABLE_STRUCTURE

Approved by: Chris Powers <chris.powers@oracle.com>
Approved by: Georgi Kodinov <georgi.kodinov@oracle.com>

marcalff committed:
WL#14611 BACKPORT SHOW PROCESSLIST TO 5.7

This patch is for 5.7 only,
to null-merge in 8.0

BACKPORT SHOW PROCESSLIST TO 5.7

Approved by: Chris Powers <chris.powers@oracle.com>

marcalff committed:
WL#14611 BACKPORT SHOW PROCESSLIST TO 5.7

Improve test robustness for:
- processlist_anonymous
- processlist_reg_user

Approved by: Chris Powers <chris.powers@oracle.com>

Commit:
Bug #34017614 Add support for Apple M1 to MySQL 5.7 [32386050]

backport 99ebfcc42dce2e656d1a76694ad5098222507777
Bug #32386050 PORT MYSQL TO THE M1 APPLE CPU/PLATFORM

Adapt to MacOS 11 on ARM

Homebrew (which we use by default for SSL) has a different location.
Use "/opt/homebrew" rather than "/usr/local".

Set the cmake variable APPLE_ARM to 1.

Set machine type to "arm64" rather than "x86_64".
Note that OS version is *not* correct, that would require backport of
Bug #32349239 WRONG VERSION REPORTED FROM MYSQLD WHEN BUILDING ON MACOS 11

Disable a unit test in non-debug mode.

Change-Id: I20551beaed36e18c1421132b0e5c9468f0287e40

Commit:
Bug #34015600 Upgrade zlib to 1.2.12 [add tarball]

Unpack source tarball, git add everything.

Change-Id: Ide09141b2a2cca0185b4c91c47496eac0fd0da1d
(cherry picked from commit 8a3f40bac9f7e874017385f834a72aa079f074a0)
(cherry picked from commit e65bb7f62df89f1653d16d91ceac1f31e956881f)

Commit:
Bug #34015600 Upgrade zlib to 1.2.12 [remove unneeded]

Remove things we do not need/want.

git rm -rf amiga/ contrib/ doc/ examples/ nintendods/ zconf.h

Change-Id: If9603ea1b8ff0f66c9396e1ab60154bb680e5e02
(cherry picked from commit 00843d74c12d1a88e42e1094b05dd2da45890bed)
(cherry picked from commit 3618a0aef8d5f113df47c09154ae605c072e633b)

Commit:
Bug #34015600 Upgrade zlib to 1.2.12 [cmake changes]

Bump MIN_ZLIB_VERSION_REQUIRED to "1.2.12"
and adjust paths to bundled zlib sources.

In extra/zlib/zlib-1.2.12/CMakeLists.txt:
 - apply cumulative patches from previous zlib upgrade
 - disable cmake code we do not want/need

This is a backport from 8.0 to 5.7.

The patch for 5.7 is copied mostly verbatim from 8.0, adaptions:
 - Remove INCLUDE_DIRECTORIES(...) from the top level CMakeLists.txt
 - If system zlib is not found, or too old, then switch to "bundled"
 - Explicitly set HAVE_COMPRESS to ON.
 - Remove "-Wmissing-profile" compiler option.
 - Remove "visibility=hidden" compiler option.

Change-Id: Iac7bed865251bf6b385f016e68e60a0dd95c5919
(cherry picked from commit ad1223f07bf172e4bf8e4e5b7ddbb01773e6e2ee)
(cherry picked from commit eb2089196d1ca7e28ec7a7e9df5e21bbe373b754)

Commit:
Bug #34017614 Add support for Apple M1 to MySQL 5.7 [WL#12928]

Backport d02d37ef9ceceebd6ac3f69dc2a776e87d336764
WL#12928 Add ARM support to MySQL Cluster

Note that this uses std::atomic_thread_fence which requires
(at least) -std=c++11
We normally build with -std=c++03 for MySQL 5.7, but not on macOS.
By not specifying any -std=<standard> we get whatever is default,
which is currently gnu++14.

Change-Id: I43f54151e04e2e856c774f8014e379905cc38d55

Commit:
Bug #34017614 Add support for Apple M1 to MySQL 5.7 [33589359]

Backport 1576e9e494fd766378a1b5bde92f665fdb4b34c5
BUG #33589359 Fix PB2 release build issues on Apple ARM

Added missing read memory barrier, this should help to control memory
access on Apple ARM platform.

Change-Id: Idf5b803c6e5c4ad30deb76f3b83ecf586fc21fbf

Commit:
Bug #34017614 Add support for Apple M1 to MySQL 5.7 [33559219]

Backport 0ace2aa267f72460bf9c0a978ab38c5b9bb3888c
BUG #33559219 Crash in pthread_cond_signal on bus error on Mac OS X ARM

Moved NdbCondition_Signal into the mutex context to avoid the condition
being destroyed during execution of NdbCondition_Signal.

Change-Id: I8e6e7c59c8ab34f12c810eb158d54eba13b88702

MRajcic committed:
Bug#32442749: SET @MASTER_BINLOG_CHECKSUM= NULL

Description:
-----------
Server exits when handling binlog dump command (COM_BINLOG_DUMP)
if the master binlog checksum algorithm was previously set to NULL
(SET @MASTER_BINLOG_CHECKSUM=NULL) on same connection.

Analysis:
--------
When binlog dump is requested by replica, server initializes
master binlog checksum algorithm from system variable, and will
exit if the respective system variable was set to NULL, because
the code will try to dereference NULL pointer in the process.

Fix:
----
Check pointer value for NULL before using it.

Approved by: Michal Jankowski <michal.jankowski@oracle.com>
Approved by: Neha Kumari <neha.n.kumari@oracle.com>

Change-Id: I01345138b1605165e0bca494133b18f582970672

gkodinov committed:
WL#14547: Update the AWS SDK used for keyring_aws
RB#27752
the 5.7 version

Commit:
Bug #34052301 MySQL 5.7: Add missing COMPONENT to INFO_BIN and INFO_SRC

backport to 5.7
Bug #29999316 ADD MISSING COMPONENT TO INSTALLED FILES

    These had Unspecified component when installed:
    docs/INFO_BIN
    docs/INFO_SRC

    Set COMPONENT Info when installing them.

Bug #33972317 Server 8.0 MSI is missing INFO_BIN and INFO_SRC

    In pushbuild we have *two* copies of the file INFO_SRC

     - ${CMAKE_SOURCE_DIR}/Docs/INFO_SRC
       This is created when the source tarball is created.
       It will be installed with this cmake code:
       INSTALL(DIRECTORY Docs/ DESTINATION ${INSTALL_DOCDIR}

     - ${CMAKE_BINARY_DIR}/Docs/INFO_SRC
       This is created during cmake configure/build.
       It will be installed with this cmake code:
       INSTALL(FILES
         ${CMAKE_BINARY_DIR}/Docs/INFO_SRC
         ${CMAKE_BINARY_DIR}/Docs/INFO_BIN
         DESTINATION ${INSTALL_DOCDIR} COMPONENT Info

    The result is that we have two entries with the same name,
    and creation of MSI or ZIP builds on Windows will fail.

    ${CMAKE_BINARY_DIR}/Docs/INFO_SRC is removed by build scripts in
    pushbuild, thus avoiding the file duplication.

    This patch removes ${CMAKE_SOURCE_DIR}/Docs/INFO_SRC
    before building the MSI target.

    With this removal in place, we can add the 'Info' COMPONENT
    to MSI builds on Windows, which is what this bug# was really about.

Change-Id: I6916d1cf826fa77e713ed9537c5108609205d897

Commit:
Bug#31933415 : OUT OF BOUND READ IN ER().

Backport. Fixed the case of unknown error.

RB#27748

Commit:
Merge branch 'mysql-5.6' into mysql-5.7

gkodinov committed:
Bug #32101184: INNODB.IMPORT_COMPRESS_ENCRYPT FAILS ON 32 BUT WITH KEYRING LOAD ERROR

The keyring_file format is not portable in 5.7.
Thus, splitting the test that relies on stored keyring data files into
two: one for all platforms and a new one that's using the stored data
and is to run only on 64 bit platforms.

Change-Id: Ic431252ee1a20cf2276f85067a5ad71bebbe0453

marcalff committed:
Bug#106939 TRUNCATE performance_schema.accounts causes duplicated counts in global_status

Contribution by Tencent.

Problem:
========

The statement

  TRUNCATE performance_schema.accounts;

can cause statistics reported by:

  SHOW GLOBAL STATUS
  SELECT * FROM performance_schema.global_status;

to be incorrect, with session status variables counted twice.

This happens only when:
- show_compatibility_56 is 0 (so that status variables
  are aggregated by accounts, users and hosts)
- some hosts are not instrumented,
  for example with a low performance-schema-hosts-size.

Root cause:
===========

Session statistics are aggregated already
from each THD (THD::status_var)
to the global status variable (global_status_var)

This is expected.

The code path for TRUNCATE performance_schema.accounts,
in function PFS_account::aggregate_status(),
contains an extra aggregation to global_status_var,
following the design pattern of other instruments.

This aggregation is not needed in case of status variables,
because of the THD -> global_status_var flow
already implemented in the SQL layer.

Fix:
====

Simplify PFS_account::aggregate_status(),
never aggregate to global_status_var.

Change-Id: Id1a83621bd6c0ff998f95041cd929dc478ab7362

Commit:
BUG#33764808: UPDATE COPYRIGHT HEADERS TO 2022 [5.7 patch]

Update all copyright headers in 5.7 to 2022 and to the current
copyright format.  This patch was generated by a script.

Approved-by: erlend.dahl@oracle.com
Change-Id: I4e4bdf6fec8240e32c6c86acbc051f136824abb7

Commit:
BUG#33764808: UPDATE COPYRIGHT HEADERS TO 2022 [5.7 patch]

Post-push fix: re-record a .result file.

Change-Id: I8f6602cfef1b9437067826fbfca6ac7b64969622

Commit:
Bug #34015600 Upgrade zlib to 1.2.12 [remove old]

Remove all old source files.

Change-Id: I42e914b4f71c40723aa35f8bd156299ce4c585a3

marcalff committed:
Bug#106939 TRUNCATE performance_schema.accounts causes duplicated counts in global_status

Post merge fix, 5.7 only.

Disable new test perfschema.misc_global_status when using the query cache.

Approved by: Georgi Kodinov <georgi.kodinov@oracle.com>

Commit:
Bug #27325012 : SPACE MISSING IN OOS ERROR MESSAGE.

Fix :
- Adds the missing space in error message.

Reviewed by : Sachin Z Agarwal (sachin.z.agarwal@oracle.com)
RB #27260

gkodinov committed:
Addendum to the fix for WL#14547: suppressed valgrind leak reports in
AWS SDK
Approved by: Harin Vadodaria <harin.vadodaria@oracle.com>

nacarvalho committed:
BUG#34144531: Backport BUG#34095747 to mysql-5.7

Replication does use write-sets to detect dependencies and/or
conflicts among transactions.
The write-set is extracted from primary, not null unique and
not null foreign keys when the binary log format is ROW and the
option `transaction_write_set_extraction` is enabled, both options
are enabled by default.

It was discovered that multi-column foreign key write-sets were not
properly generated, more precisely it was considering that foreign
keys could only refer single column, while it can also refer
multi-column keys.

In order to solve the above issue, now the foreign key write-sets do
include all referenced key columns.

Change-Id: Iee007f091c866bba9aa32d7a63f46751d63ebcb3

Commit:
BUG #28934351 : INNODB: THE TABLE IS FULL ERROR ON INSERT
BEYOND 4GB IN 32-BIT

Problem:
- Cannot increase table size beyond 4GB on 32-bit build.

Fix:
- Use uint64_t instead of ulint where overflow was
  happening on reaching 4GB mark.

Change-Id: I92d5e930d863743214f1cf5a17191128892f4f03

bkandasa committed:
Raise version number after cloning 5.6.57

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

bkandasa committed:
Update License Book

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

bkandasa committed:
Raise version number after cloning 5.7.39

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

Commit:
Bug#34144951: InnoDB restore position crash when descent scan btree

- This assertion error is only on debug build and no problems
  for normal build. It is from debug code problem.
- btr_cur_optimistic_latch_leaves() get previous block in optimistic,
  and reconfirm page_ids after the current block got latched also.
- But btr_block_get() is intended only for certain use, not for optimistic.
  And it has rigid assertion code.

So, btr_cur_optimistic_latch_leaves() should call buf_page_get_gen()
directly with BUF_GET_POSSIBLY_FREED option, to avoid the rigid assertion.

Change-Id: I05b151d627acf66301ed1c8c081e801af15c1813

sreedhars committed:
BUG#34299825 Server 5.7 - Windows msi installation requires Visual Studio 2019 x64 Redistributable
    As we recently upgraded to use newer version of Visualstudio.

Change-Id: I14b8192d6a3f035f5c988641fe700ffa97c8f7e6

Commit:
Bug#32912868: dict_table_x_lock_indexes() can cause deadlock with concurrent purge operations

Confirmed carefully also other possibility with index latch order and implicit latch order.
This bug has the both 2 aspects.

[A] Indexes latch order in same table is [secondary --> primary]
 We should be along with description in sync0type.h.
 The current dict_table_x_lock_indexes() violates the order.
 Better to be fixed.

[B] Even if dict_table_x_lock_indexes() is atomic operation,
 still might cause deadlock with
 "[secondary index page]-->[primary index latch]-->[primary index page]"
 operation. (if dict_table_x_lock_indexes() caller will access to [secondary index page].)
 That seems to be allowed order, and included to purge and rollback paths.
 So, dict_table_x_lock_indexes() should exclude other threads' concurrent accesses.
 - User threads: already excluded by mysqld MDL
 - Purge: ## needs to stop during getting latch ##
 - Rollback after recovery: ## needed to be ended already ##

Fixing [B] is enough to fix this bug.
But also [A] is better to be fixed for future fail-tolerance.

(for 5.7 only)
"TRUNCATE TABLE" path also calls dict_table_x_lock_indexes(), and fixed.

Change-Id: I0673af125aef2f4cc14d8b86a3816c71c5becc8a

Commit:
Bug#33876690 - Server may exit when creating views with a chain
of dependencies

Issue:
The stack might get exhausted while creating a nested view which
in turn references a nested view.

This is because the SELECT nesting level limit of 64 is not
enforced while preparing a view.

Fix:
Check the nesting level while preparing a view and throw an
error if limit exceedes.

Change-Id: I4eef53af614d6c2be337e2cba32cdd27c39487ad

salmankhan123417 committed:
BUG#34159579 Bug 25203933 is still present in 5.7

Description: Whenever the grant option privilege was given to an
user, User will be able to list and change the databases.

Analysis: We had been treating the 'GRANT OPTION' privilege same
as other database related privileges ( e.g. SELECT, INSERT ,
UPDATE, EVENT etc) at all three  global, db and table levels.
Hence, this bug was observed.

Fix: This is a backport from 8.0 to 5.7.
The issue is fixed by redefining the ACL macros such that all the
db related operations are separated from GRANT_ACL. And allowing
<SHOW> and <USE> operations only if db operation privileges are
given.

Change-Id: If32e86d8d87bd0e48ccea690b67eab11901ff915

Commit:
WL #15188 Add wrappers for encryption/decryption algorithms.

Added approved key derivation functions (KDF) pbkdf2_hmac and hkdf.
This enables better key derivations in functions AES_ENCRYPT and AES_DECRYPT.
And this results into better encryption.

rb: 16601
Change-Id: I8dc24b41aaf5e59bd8c1a183a6ab32b0d8c7bd60

GBuella committed:
BUG#34380370 truncate table not free mutex in some case

release mutex on all return paths in truncate_t::truncate

Change-Id: I5aeee262503db191ced7773d2456bd32d7007c53

Commit:
Bug#33866103 : MYSQLPUMP [ERROR] (1356) WITH VIEW CONTAINING FROM
               SUBQUERY - REGRESSION

Description:
Mysqlpump may not work in cases where derived tables are generated.
This is due to privileges not being set for those tables.

Fix:
Backport. Derived tables handled separately and privileges set.

Gerrit#16933

Change-Id: Ia31d479ed2d7aba47993d2e646efb84651ee784a

karthik-kamath committed:
BUG#32893161: BACKPORT 28632725 TO 5.7

When using --log-timestamps=system, ISO8601 timestamps in
logging would always show the timezone's default offset to
UTC, even when DST was active.

Note:
This is a backport of the fix for Bug#28632725 to 5.7.

Change-Id: I488f3468cc320de555cea7bb6f911e9c2f82e225

TharunSai committed:
Bug #34094706 : require_secure_transport causes abrupt server exit when checking
                thd->active_vio

Description:
------------
When require_secure_transport is enabled, the server checks the thd->active_vio
pointer in acl_authenticate(). However, this check is not thread-safe as server
can exit abruptly if the connection is killed before checking active_vio.

Fix:
----
Provided mutex protection in the acl_authenticate() method to avoid this problem
and added a nullptr check as well for thd->active_vio.

Change-Id: I42c284dcaf3c988734010f9faf02f9c023dda90d

bkandasa committed:
Update License Book

Approved-by: Bjorn Munch <bjorn.munch@oracle.com>

VarunNagaraju committed:
PXB-2990: Merge MySQL-5.7.40

https://jira.percona.com/browse/PXB-2900

Merge tag 'mysql-5.7.40' into pxb-2.4-5.7.40-merge-2990

Conflicts:
   both modified:   CMakeLists.txt
   both modified:   storage/innobase/fil/fil0fil.cc

VarunNagaraju committed:
PXB-2990: Merge MySQL-5.7.40 (Resolved conflicts)

https://jira.percona.com/browse/PXB-2900

Fixed conflicts in CMakeLists.txt by removing the docs installation
from upstream and in fil0fil.cc by preserving a NULL check from
upstream.

Also, bumped the version.

it-percona-cla commented:
[![CLA assistant check](https://cla-assistant.percona.com:443/pull/badge/not_signed)](https://cla-assistant.percona.com:443/percona/percona-xtrabackup?pullRequest=1380) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.percona.com:443/percona/percona-xtrabackup?pullRequest=1380) before we can accept your contribution.<br/>**1** out of **10** committers have signed the CLA.<br/><br/>:white_check_mark: VarunNagaraju<br/>:x: Tor Didriksen<br/>:x: nacarvalho<br/>:x: gkodinov<br/>:x: sreedhars<br/>:x: marcalff<br/>:x: bkandasa<br/>:x: karthik-kamath<br/>:x: salmankhan123417<br/>:x: GBuella<br/><hr/>**Tor Didriksen** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.percona.com:443/check/percona/percona-xtrabackup?pullRequest=1380) it.</sub>

VarunNagaraju committed:
PXB-2990: Merge MySQL-5.7.40 (Fixed compilation error)

https://jira.percona.com/browse/PXB-2900

Fixed a compilation error in the definition of the method,
mysql_thread_set_peer_port().

rahulmalik87 commented on include/mysql/psi/mysql_thread.h:
```diff
@@ -1320,15 +1320,15 @@ static inline void inline_mysql_thread_set_psi_THD(THD *thd)
 }
 #endif /* __cplusplus */
 
+#endif
```
can you comment that is closer of which if

altmannmarcelo replied:
Why we are moving up the enclose ifdef HAVE_PSI_THREAD_INTERFACE ?

PXB should have it as false and do not compile this part of code.

Also, please include in the commit message which compilation error you were getting that lead to this fix.

rahulmalik87 reviewed (changes requested):
Please test in PXB jenkins as well.

altmannmarcelo commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -1561,17 +1561,10 @@ fil_space_is_being_truncated(
 	mutex_enter(&fil_system->mutex);
-<<<<<<< HEAD
-	fil_space_t* space = fil_space_get_by_id(id);
-	mark_for_truncate = space ? space->is_being_truncated : false;
```
we already had this fix for 5 years :D

VarunNagaraju committed:
PXB-2990: Merge MySQL-5.7.40 (Fixed test failures)

https://jira.percona.com/browse/PXB-2900

* Fixes xbcloud.sh test to check if XBCLOUD_CREDENTIALS is empty
  and skip the test.
* Fixes bootstrap.sh to download the correct tar file of the server
  for xtradb57 variants.

rahulmalik87 commented:
Please paste Jenkin's result  link

VarunNagaraju commented:
> Please paste Jenkin's result link

https://pxb.cd.percona.com/view/all/job/percona-xtrabackup-2.4-param/135/

VarunNagaraju commented:
Jenkins results: https://pxb.cd.percona.com/job/percona-xtrabackup-2.4-param/139

rahulmalik87 reviewed (approved):
LGTM

altmannmarcelo reviewed (approved):
LGTM
